### PR TITLE
fix #4697, adds backslashes quotation for exec

### LINKED ISSUE #4697: i3-dmenu-desktop can't launch electrum
## I'm submitting a…

<pre>
[x] Bug
</pre>

## Current Behavior

```
% i3-dmenu-desktop        
ERROR: Your command: exec  "sh -c \"PATH=\\"\\$HOME/.local/bin:\\$PATH\\"; electrum \""
ERROR:                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR: Expected one of these tokens: <end>, '[', 'move', 'exec', 'exit', 'restart', 'reload', 'shmlog', 'debuglog', 'border', 'layout', 'append_layout', 'workspace', 'focus', 'kill', 'open', 'fullscreen', 'sticky', 'split', 'floating', 'mark', 'unmark', 'resize', 'rename', 'nop', 'scratchpad', 'swap', 'title_format', 'title_window_icon', 'mode', 'bar'
[{"success":true},{"success":false,"parse_error":true,"error":"Expected one of these tokens: <end>, '[', 'move', 'exec', 'exit', 'restart', 'reload', 'shmlog', 'debuglog', 'border', 'layout', 'append_layout', 'workspace', 'focus', 'kill', 'open', 'fullscreen', 'sticky', 'split', 'floating', 'mark', 'unmark', 'resize', 'rename', 'nop', 'scratchpad', 'swap', 'title_format', 'title_window_icon', 'mode', 'bar'","input":"exec  \"sh -c \\\"PATH=\\\\\"\\\\$HOME/.local/bin:\\\\$PATH\\\\\"; electrum \\\"\"","errorposition":"                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^"}]
Could not launch i3-msg: 512 at /usr/bin/i3-dmenu-desktop line 493, <GEN1> line 1.
zsh: exit 2     i3-dmenu-desktop
```

## Expected Behavior

The same as `gtk-launch electrum` and wofi: should work

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 
Binary i3 version:  4.20 (2021-10-19) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.20 (2021-10-19) (pid 7630)
Loaded i3 config:
  /home/sergio/.config/i3/config (main) (last modified: Fri 26 Nov 2021 18:35:33 MSK, 488 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
</pre>

```
% grep Exec /usr/share/applications/electrum.desktop 

Exec=sh -c "PATH=\"\\$HOME/.local/bin:\\$PATH\"; electrum %u"
Exec=sh -c "PATH=\"\\$HOME/.local/bin:\\$PATH\"; electrum --testnet %u"
```

<details><summary>Config file</summary><pre>
</pre>
</details>

<pre>
Logfile URL:
</pre>

<pre>
- Linux Distribution & Version: debian sid
- Are you using a compositor (e.g., xcompmgr or compton): no
</pre>

### PR DESCRIPTION
**Pull request by 532910**
Signed-off-by: Sergio E. Nemirowski <sergio@outerface.net>

### CODE REVIEW

File: i3-dmenu-desktop [hunk 1]
```diff
@@ -482,7 +482,8 @@ EOT
     # double quote which is NOT preceded by a backslash (\).
     #
     # Therefore, we escape all double quotes (") by replacing them with \"
-    $exec =~ s/"/\\"/g;
+    $exec =~ s/\\\"/\\\\\\"/g;
```
Comment (stapelberg) [inline i3-dmenu-desktop]: Does the double quote really need to be escaped? I.e., wouldn’t `s/\\"/…` work as well?

Comment (532910) [reply i3-dmenu-desktop:485]: https://github.com/i3/i3/issues/4697#issuecomment-980901026

Comment (532910) [reply i3-dmenu-desktop]: Indeed, it doesn't!

Comment (stapelberg) [reply i3-dmenu-desktop]: I wrote `s/\\"/…` and you seemed to agree, but your pull request still reads `s/\\\"/…`?

Comment (532910) [reply i3-dmenu-desktop]: forgot to push, shame on me :|

File: i3-dmenu-desktop [hunk 2]
```diff
@@ -482,7 +482,8 @@ EOT
     # double quote which is NOT preceded by a backslash (\).
     #
     # Therefore, we escape all double quotes (") by replacing them with \"
-    $exec =~ s/"/\\"/g;
+    $exec =~ s/\\"/\\\\\\"/g;
```
Comment (stapelberg) [inline i3-dmenu-desktop:485]: So this line now replaces 1 `\` with 3 `\`.

I would have expected replacing 1 `\` with 2 `\` to be sufficient, i.e. `s/\\"/\\\\"/g;`

Can you check if 2 `\` works, too?

Comment (532910) [reply i3-dmenu-desktop:485]: This line replaces `\"` with `\\\"`, in other words it prepends backslash-quotes with two backslashes more.

Comment (532910) [reply i3-dmenu-desktop:485]:
```
% cat ~/.local/share/applications/test.desktop 
[Desktop Entry]
Exec=sh -c "PATH=\"\\$HOME/.local/bin:\\$PATH\"; xterm %u"
Name=Test
Terminal=false
Type=Application
```
This code:
```
print $exec . "\n";
$exec =~ s/\\"/\\\\\\"/g;
$exec =~ s/([^\\])"/$1\\"/g;
print $exec . "\n";
```
prints:
```
sh -c "PATH=\"\\$HOME/.local/bin:\\$PATH\"; xterm "
sh -c \"PATH=\\\"\\$HOME/.local/bin:\\$PATH\\\"; xterm \"
```

this code:
```
print $exec . "\n";
$exec =~ s/\\"/\\\\"/g;
$exec =~ s/([^\\])"/$1\\"/g;
print $exec . "\n";
```
prints:
```
sh -c "PATH=\"\\$HOME/.local/bin:\\$PATH\"; xterm "
sh -c \"PATH=\\"\\$HOME/.local/bin:\\$PATH\\"; xterm \"
ERROR: Your command: exec  "sh -c \"PATH=\\"\\$HOME/.local/bin:\\$PATH\\"; xterm \""
ERROR:                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (stapelberg) [reply i3-dmenu-desktop:485]: Got it, thanks for checking!

**Review by stapelberg — APPROVED** [review]

**Pull request merged**